### PR TITLE
allow DataSetSerialiser in-place updates & fixes missing meta-data copying

### PR DESCRIPTION
**Pull request by RalphSteinhagen**

### CODE REVIEW

**Comment by RalphSteinhagen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Coverage increased per file
===========================
+ chartfx-dataset/src/main/java/de/gsi/dataset/spi/utils/MathUtils.java  7
+ chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/iobuffer/DataSetSerialiser.java  1
+ chartfx-dataset/src/main/java/de/gsi/dataset/spi/AbstractDataSet.java  2
+ chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/iobuffer/FieldDataSetHelper.java  4
+ chartfx-dataset/src/main/java/de/gsi/dataset/spi/DefaultAxisDescription.java  2
         

Coverage decreased per file
===========================
- chartfx-dataset/src/main/java/de/gsi/dataset/utils/CacheCollection.java  -3
- chartfx-dataset/src/main/java/de/gsi/dataset/utils/DoubleArrayCache.java  -4
         

Complexity increasing per file
==============================
- chartfx-dataset/src/test/java/de/gsi/dataset/serializer/spi/iobuffer/DataSetSerialiserBenchmark.java  1
         

Complexity decreasing per file
==============================
+ chartfx-dataset/src/main/java/de/gsi/dataset/spi/DefaultAxisDescription.java  -6
         

Clones added
============
- chartfx-dataset/src/test/java/de/gsi/dataset/serializer/spi/iobuffer/DataSetSerialiserTests.java  2
         

Clones removed
==============
+ chartfx-dataset/src/main/java/de/gsi/dataset/spi/FloatDataSet.java  -3
+ chartfx-dataset/src/main/java/de/gsi/dataset/spi/DoubleErrorDataSet.java  -3
+ chartfx-dataset/src/main/java/de/gsi/dataset/spi/DoubleDataSet.java  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6127189&bid=20042346)

File: chartfx-dataset/src/test/java/de/gsi/dataset/serializer/spi/iobuffer/DataSetSerialiserBenchmark.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+import de.gsi.dataset.testdata.spi.GaussFunction;
+
+@State(Scope.Benchmark)
+public class DataSetSerialiserBenchmark {
+    private static DataSet srcDataSet = new DoubleDataSet(new GaussFunction("Gauss-function", 10_000));
+    private static DataSet copyDataSet = new DoubleDataSet(srcDataSet);
+    private static TestClass source = new TestClass();
+    private static TestClass copy = new TestClass();
+    private static IoBuffer fastByteBuffer = new FastByteBuffer(100_000);
+    private static IoBuffer byteBuffer = new ByteBuffer(100_000);
```
Comment (RalphSteinhagen) [inline chartfx-dataset/src/test/java/de/gsi/dataset/serializer/spi/iobuffer/DataSetSerialiserBenchmark.java:26]: Codacy found an issue: [Avoid unused private fields such as 'byteBuffer'.](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6127189)

**Comment by RalphSteinhagen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- chartfx-dataset/src/test/java/de/gsi/dataset/serializer/spi/iobuffer/DataSetSerialiserBenchmark.java  1
         

Complexity decreasing per file
==============================
+ chartfx-dataset/src/main/java/de/gsi/dataset/spi/DefaultAxisDescription.java  -6
         

Clones added
============
- chartfx-dataset/src/test/java/de/gsi/dataset/serializer/spi/iobuffer/DataSetSerialiserTests.java  2
         

Clones removed
==============
+ chartfx-dataset/src/main/java/de/gsi/dataset/spi/DoubleErrorDataSet.java  -3
+ chartfx-dataset/src/main/java/de/gsi/dataset/spi/FloatDataSet.java  -3
+ chartfx-dataset/src/main/java/de/gsi/dataset/spi/DoubleDataSet.java  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6127189&bid=20042346)

File: chartfx-dataset/src/test/java/de/gsi/dataset/serializer/spi/iobuffer/DataSetSerialiserBenchmark.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    private static IoClassSerialiser serialiserFastByteBuffer = new IoClassSerialiser(fastByteBuffer, BinarySerialiser.class);
+    private static IoClassSerialiser serialiserByteBuffer = new IoClassSerialiser(fastByteBuffer, BinarySerialiser.class);
```
Comment (RalphSteinhagen) [inline chartfx-dataset/src/test/java/de/gsi/dataset/serializer/spi/iobuffer/DataSetSerialiserBenchmark.java:39]: Codacy found an issue: [Avoid unused private fields such as 'serialiserByteBuffer'.](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6127189)

File: chartfx-dataset/src/test/java/de/gsi/dataset/serializer/spi/iobuffer/DataSetSerialiserBenchmark.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+ * @author rstein
+ */
+@State(Scope.Benchmark)
+public class DataSetSerialiserBenchmark {
+    private static DataSet srcDataSet = new DoubleDataSet(new GaussFunction("Gauss-function", 10_000));
+    private static DataSet copyDataSet = new DoubleDataSet(srcDataSet);
+    private static TestClass source = new TestClass();
+    private static TestClass copy = new TestClass();
+    private static IoBuffer fastByteBuffer = new FastByteBuffer(100_000);
+    private static IoBuffer byteBuffer = new ByteBuffer(100_000);
```
Comment (RalphSteinhagen) [inline chartfx-dataset/src/test/java/de/gsi/dataset/serializer/spi/iobuffer/DataSetSerialiserBenchmark.java:37]: Codacy found an issue: [Avoid unused private fields such as 'byteBuffer'.](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6127189)

**Comment by RalphSteinhagen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- chartfx-dataset/src/test/java/de/gsi/dataset/serializer/spi/iobuffer/DataSetSerialiserBenchmark.java  1
         

Complexity decreasing per file
==============================
+ chartfx-dataset/src/main/java/de/gsi/dataset/spi/DefaultAxisDescription.java  -6
         

Clones added
============
- chartfx-dataset/src/test/java/de/gsi/dataset/serializer/spi/iobuffer/DataSetSerialiserTests.java  2
         

Clones removed
==============
+ chartfx-dataset/src/main/java/de/gsi/dataset/spi/DoubleDataSet.java  -3
+ chartfx-dataset/src/main/java/de/gsi/dataset/spi/FloatDataSet.java  -3
+ chartfx-dataset/src/main/java/de/gsi/dataset/spi/DoubleErrorDataSet.java  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6127189&bid=20042346)

**Review by wirew0rm — APPROVED** [review]
Looks good :+1:

**Pull request merged**